Cross-correlation function for analog signals

mvonpapen opened the pull request:
Added a function "cross_correlation_function" in signal_processing.py that calculates the pairwise cross-correlation of selected channels within a neo.AnalogSignal. Optionally calculates the envelope of the cross-correlation function, which is helpful to determine the correlation length of an oscillatory signal.

The two signals, e.g. LFP, are zero padded to minimize boundary effects, transformed into Fourier domain, multiplied and transformed back into time domain. The signals are afterwards corrected (unbiased estimator) for zero-padding.

This function should not be confused with cross_corr_coeff option in cross_correlation_histogram function within spiketrain_correlation.py. The latter only works for CCH obtained from spiketrains and assumes Poissonian statistics (Equation (5.10) in "Analysis of parallel spike trains", 2010, Gruen & Rotter, Vol 7).

The input is 
- signal: neo.AnalogSignal
- channel1/2: int, list, tuple, array; this determines the indices of the channels; cross-correlation will be calculated for pairs channel1[0] vs. channel2[0], channel1[1] vs channel2[1], and so forth
- dt: sample resolution
- env: bool; envelope yes or no
- nlags: int; how many lags the output rho should have

The output is:
- rho: array, the cross-correation function (ranging between -1 and 1) OR env, the envelope of rho
- tau: array, vector of lags

pep8speaks commented:
Hello @mvonpapen! Thanks for submitting the PR.

 - In the file [`elephant/signal_processing.py`](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/signal_processing.py), following are the PEP8 issues :

> [Line 73:80](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/signal_processing.py#L73): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)
> [Line 137:1](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/signal_processing.py#L137): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 138:1](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/signal_processing.py#L138): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 139:80](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/signal_processing.py#L139): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (89 > 79 characters)
> [Line 143:1](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/signal_processing.py#L143): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 144:72](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/signal_processing.py#L144): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 147:72](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/signal_processing.py#L147): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 149:77](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/signal_processing.py#L149): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 164:80](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/signal_processing.py#L164): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (89 > 79 characters)
> [Line 169:69](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/signal_processing.py#L169): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 171:1](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/signal_processing.py#L171): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 181:67](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/signal_processing.py#L181): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 183:80](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/signal_processing.py#L183): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (104 > 79 characters)
> [Line 184:80](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/signal_processing.py#L184): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (112 > 79 characters)
> [Line 189:1](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/signal_processing.py#L189): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 193:1](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/signal_processing.py#L193): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 195:80](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/signal_processing.py#L195): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (91 > 79 characters)
> [Line 197:80](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/signal_processing.py#L197): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (91 > 79 characters)
> [Line 198:1](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/signal_processing.py#L198): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 202:15](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/signal_processing.py#L202): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 203:15](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/signal_processing.py#L203): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 204:1](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/signal_processing.py#L204): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 208:1](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/signal_processing.py#L208): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 210:74](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/signal_processing.py#L210): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 214:16](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/signal_processing.py#L214): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 214:50](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/signal_processing.py#L214): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 214:61](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/signal_processing.py#L214): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 217:1](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/signal_processing.py#L217): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 222:20](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/signal_processing.py#L222): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 222:33](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/signal_processing.py#L222): [E201](https://duckduckgo.com/?q=pep8%20E201) whitespace after '('
> [Line 222:62](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/signal_processing.py#L222): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 222:66](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/signal_processing.py#L222): [E202](https://duckduckgo.com/?q=pep8%20E202) whitespace before ')'
> [Line 223:1](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/signal_processing.py#L223): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 227:33](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/signal_processing.py#L227): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 228:42](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/signal_processing.py#L228): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 230:1](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/signal_processing.py#L230): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 330:80](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/signal_processing.py#L330): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)

 - In the file [`elephant/test/test_signal_processing.py`](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/test/test_signal_processing.py), following are the PEP8 issues :

> [Line 29:80](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/test/test_signal_processing.py#L29): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 38:24](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/test/test_signal_processing.py#L38): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 40:16](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/test/test_signal_processing.py#L40): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 41:16](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/test/test_signal_processing.py#L41): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 42:16](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/test/test_signal_processing.py#L42): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 43:16](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/test/test_signal_processing.py#L43): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 45:71](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/test/test_signal_processing.py#L45): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 51:80](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/test/test_signal_processing.py#L51): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 79 characters)
> [Line 52:28](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/test/test_signal_processing.py#L52): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 54:44](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/test/test_signal_processing.py#L54): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 54:80](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/test/test_signal_processing.py#L54): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 55:44](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/test/test_signal_processing.py#L55): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 55:80](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/test/test_signal_processing.py#L55): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (88 > 79 characters)
> [Line 56:80](https://github.com/NeuralEnsemble/elephant/blob/8feaccd83d61679341dadef0c676df28a2edc1dc/elephant/test/test_signal_processing.py#L56): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)

JuliaSprenger commented on elephant/signal_processing.py:
```diff
@@ -134,6 +135,111 @@ def zscore(signal, inplace=True):
+        rounded to nearest integer. Output is of length 2*nlags+1. If None, 
+        len(tau) = Nt
+
+    Returns
```
The description of the return values seems to be incomplete. How about going more into details about the envelope / rho calculation. Also mention tau here.

JuliaSprenger commented on elephant/signal_processing.py:
```diff
@@ -134,6 +135,111 @@ def zscore(signal, inplace=True):
+    ch_pairs : list (or array with shape (N,2))
+        list with N channel pairs for which to compute cross-correlation, 
+        each element of list must contain 2 channel indices
+    dt : float
```
And if you use ```dt``` as a sampling time, then it should be a quantity and not a float.

JuliaSprenger replied:
```dt``` should be always larger or equal to the sampling period of the signals.

mvonpapen replied:
Good point. I will extract dt from the signal then.

JuliaSprenger commented on elephant/signal_processing.py:
```diff
@@ -134,6 +135,111 @@ def zscore(signal, inplace=True):
+    R(tau) = 1/(N-|k|) R'(tau), where R'(tau) = E[x(t)*y(t+tau)] for signal1 
+    and signal2 in a pairwise manner, i.e. signal1[0] vs signal2[0], signal1[1] 
+    vs signal2[1] and so on. The cross-correlation function is obtained 
+    by np.correlate. signal1 and signal2 are zscored beforehand.
```
You seem not to use ```np.correlate```

mvonpapen replied:
description has been changed accordingly

JuliaSprenger commented on elephant/signal_processing.py:
```diff
@@ -134,6 +135,111 @@ def zscore(signal, inplace=True):
+    env: bool
+        Return Hilbert envelope of cross-correlation function
+    nlags: int
+        Defines number of lags for cross-corelation function. Floats will be 
```
```Floats``` should read ```Float```

JuliaSprenger commented on elephant/signal_processing.py:
```diff
@@ -134,6 +135,111 @@ def zscore(signal, inplace=True):
+    nlags: int
+        Defines number of lags for cross-corelation function. Floats will be 
+        rounded to nearest integer. Output is of length 2*nlags+1. If None, 
+        len(tau) = Nt
```
What is ```Nt```? The number of samples in the signals?

mvonpapen replied:
Nt is now clearly defined

mvonpapen commented on elephant/test/test_signal_processing.py:
```diff
@@ -15,12 +15,48 @@
+            signal = neo.AnalogSignal(x, units='mV', t_start=0.*pq.ms, 
+                                      sampling_rate=1/dt*pq.Hz, dtype=float)
+            rho, tau = elephant.signal_processing.cross_correlation_function(
+                    signal, [[0, 2], [1, 3]], dt=dt, nlags=N/4.6)
```
No, it doesnt. I just wanted to use some float smaller than the number of input samples. Do you think it would be better to use a fixed float, e.g. 31.3? Or what would you suggest?

JuliaSprenger replied:
It just looks like it has some deeper meaning if you write it like this. So you should either add a comment, telling that 4.6 is just a random choice, or use a fixed float, which you can also define together with all the other parameters above.

mvonpapen replied:
I will define nlag = 30 in the beginning. I only use nteger in order not to complicate things

JuliaSprenger commented on elephant/signal_processing.py:
```diff
@@ -134,6 +135,111 @@ def zscore(signal, inplace=True):
+    Computes unbiased estimator of the cross-correlation function.
+    
+    Calculates the unbiased estimator of the cross-correlation function 
+    R(tau) = 1/(N-|k|) R'(tau), where R'(tau) = E[x(t)*y(t+tau)] for signal1 
```
Can you add a reference for this normalization? Also please add a description of the variables you are using.

mvonpapen replied:
Will add: Hall & River (2009) Spectral Analysis of Signals, Spectral Element Method in Structural Dynamics, Eq. 2.2.3

JuliaSprenger commented:
This function is accompanied by a single, quite complex test. We suggest to add multiple, more simple and stupid tests, eg like testing that ```tau``` is independent of the ```env``` keyword.

mvonpapen commented:
> This function is accompanied by a single, quite complex test. We suggest to add multiple, more simple and stupid tests, eg like testing that `tau` is independent of the `env` keyword.

I added separate tests for env, nlags, different frequencies and phase shift

alperyeg commented:
@JuliaSprenger let me know when to merge. Thank you for the review.

JuliaSprenger commented on elephant/signal_processing.py:
```diff
@@ -134,6 +135,124 @@ def zscore(signal, inplace=True):
+        Return Hilbert envelope of cross-correlation function
+        Default: False
+    nlags: int
+        Defines number of lags for cross-corelation function. Float will be 
```
'corelation' -> 'correlation'

JuliaSprenger commented on elephant/signal_processing.py:
```diff
@@ -134,6 +135,124 @@ def zscore(signal, inplace=True):
+        Default: False
+    nlags: int
+        Defines number of lags for cross-corelation function. Float will be 
+        rounded to nearest integer. Sample length of output is 2*nlags+1. 
```
'Sample length'? maybe better 'Number of samples' or similar?

JuliaSprenger commented on elephant/signal_processing.py:
```diff
@@ -134,6 +135,124 @@ def zscore(signal, inplace=True):
+        Defines number of lags for cross-corelation function. Float will be 
+        rounded to nearest integer. Sample length of output is 2*nlags+1. 
+        If None, sample length of output is equal to number of samples in 
+        input signal, namely Nt
```
Nt.

JuliaSprenger commented on elephant/signal_processing.py:
```diff
@@ -134,6 +135,124 @@ def zscore(signal, inplace=True):
+        plt.show()
+    """
+    
+    # Make ch_pairs an 2D array
```
'an' -> 'a'

JuliaSprenger commented:
Thanks for the improvements! The code is already more readable!
I would suggest to add some more checks on the input parameters. Right now the error messages for messed up inputs are not really telling. For example when providing a negative value for `nlags` or a string as `env` parameter.

mvonpapen commented:
Hi Julia,

Thanks very much for the thorouh review. I added the suggested checks for the input (keyword) arguments and corrected for pep8. I additionally changed some variable names to comply with snake_case naming style.

While doing pep8 checks, I noticed that the refractor test complained about too many local variables. I therefore also reduced the number of these where possible (e.g. variables a, b, and bias).

JuliaSprenger commented:
Ready to merge from my side! Thanks for fixing all the details!

alperyeg commented on elephant/signal_processing.py:
```diff
@@ -136,6 +137,129 @@ def zscore(signal, inplace=True):
+    Computes unbiased estimator of the cross-correlation function.
+
+    Calculates the unbiased estimator of the cross-correlation function
+    R(tau) = 1/(N-|k|) R'(tau), where R'(tau) = E[x(t)*y(t+tau)] in a
```
You could use `.. math::` for the mathematical formulations to have them in a nice format on the docs page. 
```suggestion
    .. math:: R(tau) = frac{1}/{(N-|k|)} R^{\prime}(tau), \text{where} R^{\prime}(tau) = E[x(t)*y(t+tau)] 
```
Please have a look at section 12 of the [numpy docs](https://numpydoc.readthedocs.io/en/latest/format.html) for possibilities.

mvonpapen replied:
Good idea and I tried it. Unfortunately, the latest Spyder version 3.2.6 cannot render this and the output becomes shitty (see below). I would therefore suggest not to write it like this.

--- docstring output ---

Calculates the unbiased estimator of the cross-correlation function [1]

.. math:: R(        au) = rac{1}{N-|k|} R'(        au) \ ,

ight>` in a pairwise manner,eft<x(t)y(t+    au)
i.e. signal[ch_pairs[0,0]] vs signal2[ch_pairs[0,1]], signal[ch_pairs[1,0]]
vs signal2[ch_pairs[1,1]], and so on. The cross-correlation function is
obtained by scipy.signal.fftconvolve. Time series in signal are zscored
beforehand. Alternatively returns the Hilbert envelope of :math:`R( au)`,
which is useful to determine the correlation length of oscillatory signals.

[1] Hall & River (2009) "Spectral Analysis of Signals, Spectral Element
Method in Structural Dynamics", Eq. 2.2.3

alperyeg commented on elephant/signal_processing.py:
```diff
@@ -136,6 +137,129 @@ def zscore(signal, inplace=True):
+
+    Calculates the unbiased estimator of the cross-correlation function
+    R(tau) = 1/(N-|k|) R'(tau), where R'(tau) = E[x(t)*y(t+tau)] in a
+    pairwise manner, i.e. signal[ch_pairs[0,0]] vs signal2[ch_pairs[0,1]],
```
Encapsulate variables or code snippets etc with  ` `` `
```suggestion
`signal[ch_pairs[0,0]]` vs `signal2[ch_pairs[0,1]]`
```

alperyeg commented on elephant/signal_processing.py:
```diff
@@ -136,6 +137,129 @@ def zscore(signal, inplace=True):
+    R(tau) = 1/(N-|k|) R'(tau), where R'(tau) = E[x(t)*y(t+tau)] in a
+    pairwise manner, i.e. signal[ch_pairs[0,0]] vs signal2[ch_pairs[0,1]],
+    signal[ch_pairs[1,0]] vs signal2[ch_pairs[1,1]], and so on (see also
+    Hall & River (2009) Spectral Analysis of Signals, Spectral Element Method
```
I would put this on the references section (References) and only reference it here.

alperyeg commented on elephant/signal_processing.py:
```diff
@@ -136,6 +137,129 @@ def zscore(signal, inplace=True):
+        pairwise cross-correlation function. This is helpful to compute the
+        correlation length for oscillating cross-correlation functions
+
+    Example:
```
This is not the correct format for the example section, please have a look e.g. [here](https://github.com/NeuralEnsemble/elephant/blob/master/elephant/cell_assembly_detection.py#L189)

alperyeg commented on elephant/signal_processing.py:
```diff
@@ -136,6 +137,129 @@ def zscore(signal, inplace=True):
+        pairs = pairs[:, np.newaxis]
+
+    # Check input
+    if not isinstance(signal, neo.AnalogSignal):
```
Do you have a raises section (Raises) in the documentation, you should mention there the `ValueError` and `KeyError`.

mvonpapen replied:
added raises section

alperyeg commented on elephant/signal_processing.py:
```diff
@@ -136,6 +137,129 @@ def zscore(signal, inplace=True):
+
+    # Return neo.AnalogSignal
+    cross_corr = neo.AnalogSignal(xcorr,
+                                  units='',
```
Why do you give an empty string for the `units`

mvonpapen replied:
correlation is unitless

alperyeg replied:
Right, didn't even now that this is possible in `neo`.
